### ISSUE TITLE
Built-in/efficient way for state to return the entirety of current domain

### ISSUE THREAD

**Issue by mmcilree** [ciaranm/glasgow-constraint-solver#26]
Smart Table requires this is in several places
```C++
state.for_each_value(var, [&](Integer value) {
                unsupported[var].emplace_back(value);
            });
```
...
```C++
state.for_each_value(var, [&](Integer value) {
                        supported_by_tree[var].emplace_back(value);
                    });
```
And it looks like the generic MDD membership will need to do something like this too. Is it work having a `state.current_domain_copy(var)` or something?

**Comment by ciaranm** [comment]
I've not implemented this because in the long term, I'm hoping we'll never have to explicitly store each value in a domain. If we had some kind of IntervalSet data structure that could compactly handle domains like 1..1000,1023..2000,3200..4000, would SmartTable be able to use this instead of an explicit set?

**Comment by mmcilree** [comment]
Yes, this would be good. In fact the original MDD propagator paper advises the use of sparse sets, and I think these can be used for SmartTable too (c.f. Compact Table).

**Comment by ciaranm** [comment]
I think we do in fact have this now.

**Comment by ciaranm** [comment]
Resolved. The requested whole-domain copy now exists as `State::copy_of_values(var) -> IntervalSet<Integer>` (alongside the `each_value_immutable` / `each_value_mutable` generators), and `smart_table.cc` no longer stores domains as explicit value vectors via the `for_each_value(...) { ...emplace_back(value); }` pattern this issue cited -- it works on `IntervalSet` domains through `get_for_actual_var` / `set_for_actual_var`. This matches the direction in the thread (the compact `IntervalSet` domain representation), and the last comment already noted "I think we do in fact have this now." Closing.
